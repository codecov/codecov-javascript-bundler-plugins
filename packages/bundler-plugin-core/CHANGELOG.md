# @codecov/bundler-plugin-core

## 0.0.1-beta.0

### Patch Changes

- f5cae1c: Move packages into beta

## 0.0.1-beta.0

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 543a526: Update dependencies before moving package out of alpha
- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.
- 36ed299: Alpha release of Codecov bundler plugins
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- 48b6e90: Move bundler-core test deps to dev deps
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.3

### Patch Changes

- 543a526: Update dependencies before moving package out of alpha
- c1fdbd6: Fix retry when fetching by throwing custom error if response is not okay
- d28a2e4: Break from retry loop when the response is ok
- 562ac0c: Adjust peer dep versions to use x-range instead of caret
- c755a5c: Add missing outputPath to output stats file

## 0.0.1-alpha.2

### Patch Changes

- becd728: Add duration to webpack bundler plugin directly
- 7c55993: Rename uploaderOverrides to uploadOverrides, remove url from uploadOverrides, and removal of repoToken/globalUploadToken in favour of just uploadToken.

## 0.0.1-alpha.1

### Patch Changes

- 4f1183e: Remove references to Webpack 4, as we currently on support Node 18+.
- 48b6e90: Move bundler-core test deps to dev deps

## 0.0.1-alpha.0

### Patch Changes

- 36ed299: Alpha release of Codecov bundler plugins
