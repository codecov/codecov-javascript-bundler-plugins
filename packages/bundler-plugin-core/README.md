<p align="center">
  <a href="https://about.codecov.io" target="_blank">
    <img src="https://about.codecov.io/wp-content/themes/codecov/assets/brand/sentry-cobranding/logos/codecov-by-sentry-logo.svg" alt="Codecov by Sentry logo" width="280" height="84">
  </a>
</p>

# Codecov Bundler Plugin Core

> [!WARNING]  
> These plugins are currently in beta and are subject to change.

> The bundler-plugin-core is not to be used individually and is meant to be used in conjunction with the bundler specific plugins.

Core package containing the bundler-agnostic logic for the Codecov Bundler Plugins.

Checkout the individual packages for more information and examples:

- [Rollup](https://npmjs.org/package/@codecov/rollup-plugin)
- [Vite](https://npmjs.org/package/@codecov/vite-plugin)
- [Webpack](https://npmjs.org/package/@codecov/webpack-plugin)

## Features

The Codecov Bundler Core package contains the following functionality:

- Bundle Analysis Stats Upload

## More information

- [Bundler Config Docs](https://codecov.github.io/codecov-javascript-bundler-plugins/modules/_codecov_bundler_plugin_core.html)
- [Codecov Documentation](https://docs.codecov.com/docs)
- [Codecov Feedback](https://github.com/codecov/feedback/discussions)
- [Sentry Discord](https://discord.gg/Ww9hbqr)
