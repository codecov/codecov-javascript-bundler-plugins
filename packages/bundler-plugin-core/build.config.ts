import { defineBuildConfig } from "unbuild";
import { codecovRollupPlugin } from "codecovProdRollupPlugin";
import packageJson from "./package.json";

export default defineBuildConfig({
  entries: ["./src/index"],
  outDir: "dist",
  declaration: "compatible",
  sourcemap: true,
  rollup: {
    dts: {
      compilerOptions: {
        removeComments: false,
      },
    },
    emitCJS: true,
    esbuild: {
      minify: true,
    },
  },
  hooks: {
    "rollup:options": (_ctx, opts) => {
      if (process.env.PLUGIN_CODECOV_TOKEN && Array.isArray(opts.plugins)) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        opts.plugins = [
          ...opts.plugins,
          codecovRollupPlugin({
            enableBundleAnalysis:
              typeof process.env.PLUGIN_CODECOV_TOKEN === "string",
            bundleName: packageJson.name,
            uploadToken: process.env.PLUGIN_CODECOV_TOKEN,
            apiUrl: process.env.PLUGIN_CODECOV_API_URL,
          }),
        ];
      }
    },
  },
});
