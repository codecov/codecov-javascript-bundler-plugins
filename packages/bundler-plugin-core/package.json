{
  "name": "@codecov/bundler-plugin-core",
  "version": "0.0.1-beta.3",
  "description": "Official Codecov Bundler Plugin Core",
  "author": "Codecov",
  "license": "MIT",
  "type": "module",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "type-check": "tsc --noEmit",
    "build": "unbuild",
    "dev": "unbuild --stub && node --watch-path=src dist/index.mjs",
    "clean": "rm -rf dist node_modules",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "format": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --write",
    "format:check": "prettier '**/*.{cjs,mjs,ts,tsx,md,json}' --ignore-path ../.gitignore --ignore-unknown --no-error-on-unmatched-pattern --check",
    "test:unit": "jest",
    "test:unit:ci": "jest --coverage"
  },
  "dependencies": {
    "chalk": "4.1.2",
    "semver": "^7.5.4",
    "unplugin": "^1.6.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@swc/core": "^1.3.107",
    "@swc/jest": "^0.2.33",
    "@types/jest": "^29.5.11",
    "@types/node": "^20.11.15",
    "@types/semver": "^7.5.6",
    "jest": "^29.7.0",
    "msw": "^2.1.5",
    "codecovProdRollupPlugin": "npm:@codecov/rollup-plugin@0.0.1-beta.2",
    "testdouble": "^3.20.1",
    "testdouble-jest": "^2.0.0",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3",
    "unbuild": "^2.0.0"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "engines": {
    "node": ">=18.0.0"
  }
}
