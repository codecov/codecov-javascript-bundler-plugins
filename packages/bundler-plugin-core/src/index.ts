import { type UnpluginOptions, createUnplugin } from "unplugin";
import { satisfies } from "semver";
import { z } from "zod";

import {
  type BundleAnalysisUploadPlugin,
  type Asset,
  type Chunk,
  type Module,
  type Options,
  type Output,
} from "./types.ts";

import { jsonSchema } from "./schemas.ts";
import { red } from "./utils/logging.ts";

const NODE_VERSION_RANGE = ">=18.18.0";

interface CodecovUnpluginFactoryOptions {
  bundleAnalysisUploadPlugin: BundleAnalysisUploadPlugin;
}

export function codecovUnpluginFactory({
  bundleAnalysisUploadPlugin,
}: CodecovUnpluginFactoryOptions) {
  return createUnplugin<Options, true>((userOptions, unpluginMetaContext) => {
    const plugins: UnpluginOptions[] = [];

    if (!satisfies(process.version, NODE_VERSION_RANGE)) {
      red(
        `Codecov ${unpluginMetaContext.framework} bundler plugin requires Node.js ${NODE_VERSION_RANGE}. You are using Node.js ${process.version}. Please upgrade your Node.js version.`,
      );
      process.exit(1);
    }

    if (userOptions?.enableBundleAnalysis) {
      const statsFileName = z
        .string()
        .endsWith(".json")
        .optional()
        .parse(userOptions?.statsFileName);

      const output: Output = {
        version: "1",
      };

      let startTime = NaN;
      const { pluginVersion, version, ...pluginOpts } =
        bundleAnalysisUploadPlugin({
          output,
          statsFileName,
        });

      plugins.push({
        ...pluginOpts,
        buildStart() {
          startTime = Date.now();
          output.version = version;
          output.plugin = {
            name: pluginOpts.name,
            version: pluginVersion,
          };
          output.builtAt = startTime;
        },
        buildEnd(this) {
          const duration = Date.now() - startTime;
          output.duration = duration;
        },
      });
    }

    return plugins;
  });
}

export type { BundleAnalysisUploadPlugin, Asset, Chunk, Module, Options };
export { jsonSchema };
