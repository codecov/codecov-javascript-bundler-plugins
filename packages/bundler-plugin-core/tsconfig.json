{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "@/*": ["./src/*"],
      "@test-utils/*": ["./test-utils/*"]
    },
    "checkJs": true
  },
  "include": [
    "src",
    "test-utils",
    "build.config.ts",
    "../../reset.d.ts",
    "jest.config.ts"
  ]
}
