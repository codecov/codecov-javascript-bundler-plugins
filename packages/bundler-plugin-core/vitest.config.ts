import { defineConfig } from "vitest/config";
import { config } from "../../vitest.shared";

export default defineConfig({
  ...config,
});
