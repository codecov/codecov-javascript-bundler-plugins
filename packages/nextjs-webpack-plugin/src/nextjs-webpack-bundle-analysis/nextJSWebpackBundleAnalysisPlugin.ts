import { red, type ExtendedBAUploadPlugin } from "@codecov/bundler-plugin-core";
import type * as webpack from "webpack";

import {
  _internal_processAssets as processAssets,
  _internal_processChunks as processChunks,
  _internal_processModules as processModules,
} from "@codecov/webpack-plugin";

export const nextJSWebpackBundleAnalysisPlugin: ExtendedBAUploadPlugin<{
  webpack: typeof webpack | null;
}> = ({ output, pluginName, pluginVersion }) => ({
  version: output.version,
  name: pluginName,
  pluginVersion,
  buildStart: () => {
    output.start();
    output.setPlugin(pluginName, pluginVersion);
  },
  buildEnd: () => {
    output.end();
  },
  writeBundle: async () => {
    await output.write();
  },
  webpack(compiler) {
    compiler.hooks.thisCompilation.tap(pluginName, (compilation) => {
      if (!webpack) {
        red(
          "Unable to run bundle analysis, Webpack wasn't passed successfully.",
        );
        return;
      }

      compilation.hooks.processAssets.tapPromise(
        {
          name: pluginName,
          stage: webpack.Compilation.PROCESS_ASSETS_STAGE_REPORT,
        },
        async () => {
          output.setBundleName(output.originalBundleName);
          // Webpack base chunk format options: https://webpack.js.org/configuration/output/#outputchunkformat
          if (typeof compilation.outputOptions.chunkFormat === "string") {
            if (compilation.name && compilation.name !== "") {
              output.setBundleName(`${output.bundleName}-${compilation.name}`);
            }

            let chunkFormat = compilation.outputOptions.chunkFormat;
            if (chunkFormat === "commonjs") {
              chunkFormat = "cjs";
            } else if (chunkFormat === "module") {
              chunkFormat = "esm";
            }

            output.setBundleName(`${output.bundleName}-${chunkFormat}`);
          }

          const compilationStats = compilation.getStats().toJson({
            assets: true,
            chunks: true,
            modules: true,
            builtAt: true,
            hash: true,
          });

          output.bundler = {
            name: "webpack",
            version: webpack.version,
          };

          const outputOptions = compilation.outputOptions;
          const { assets, chunks, modules } = compilationStats;

          if (assets) {
            const collectedAssets = await processAssets({
              assets,
              compilation,
            });

            output.assets = collectedAssets;
          }

          // need to collect all possible chunk ids beforehand
          // this collection is done in the processChunks function
          const chunkIdMap = new Map<number | string, string>();
          if (chunks) {
            output.chunks = processChunks({ chunks, chunkIdMap });
          }

          if (modules) {
            output.modules = processModules({ modules, chunkIdMap });
          }

          output.duration = Date.now() - (output.builtAt ?? 0);
          output.outputPath = outputOptions.path ?? "";

          // only output file if running dry run
          if (output.dryRun) {
            const { RawSource } = webpack.sources;
            compilation.emitAsset(
              `${output.bundleName}-stats.json`,
              new RawSource(output.bundleStatsToJson()),
            );
          }
        },
      );
    });
  },
});
