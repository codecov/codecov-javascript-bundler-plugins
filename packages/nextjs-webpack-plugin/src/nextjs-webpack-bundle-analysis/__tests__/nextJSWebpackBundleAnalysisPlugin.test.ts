import { Output } from "@codecov/bundler-plugin-core";
import { describe, it, expect } from "vitest";
import * as webpack from "webpack";
import { nextJSWebpackBundleAnalysisPlugin } from "../nextJSWebpackBundleAnalysisPlugin";

// @ts-expect-error this value is being replaced by rollup
const PLUGIN_NAME = __PACKAGE_NAME__ as string;
// @ts-expect-error this value is being replaced by rollup
const PLUGIN_VERSION = __PACKAGE_VERSION__ as string;

describe("webpackBundleAnalysisPlugin", () => {
  describe("when called", () => {
    it("returns a plugin object", () => {
      const plugin = nextJSWebpackBundleAnalysisPlugin({
        output: new Output({
          apiUrl: "http://localhost",
          bundleName: "test-bundle",
          debug: false,
          dryRun: true,
          enableBundleAnalysis: true,
          retryCount: 1,
          uploadToken: "test-token",
          telemetry: false,
        }),
        options: {
          webpack: webpack,
        },
        pluginName: PLUGIN_NAME,
        pluginVersion: PLUGIN_VERSION,
      });

      expect(plugin).toMatchSnapshot();
    });
  });
});
