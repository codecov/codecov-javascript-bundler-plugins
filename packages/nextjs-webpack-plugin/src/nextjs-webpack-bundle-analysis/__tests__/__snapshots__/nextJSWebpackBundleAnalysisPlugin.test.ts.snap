// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webpackBundleAnalysisPlugin > when called > returns a plugin object 1`] = `
{
  "buildEnd": [Function],
  "buildStart": [Function],
  "name": "@codecov/nextjs-webpack-plugin",
  "pluginVersion": "1.2.1",
  "version": "2",
  "webpack": [Function],
  "writeBundle": [Function],
}
`;
